### PR TITLE
Fix NEO callstates

### PR DESCRIPTION
**Pull request by shargon**
# Description

Fixes https://github.com/neo-project/neo/issues/3596

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Manual

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Neo/SmartContract/Native/NativeContract.cs [hunk 1]
```diff
@@ -379,8 +383,8 @@ internal async void Invoke(ApplicationEngine engine, byte version)
                     throw new InvalidOperationException($"Cannot call this method before hardfork {method.ActiveIn}.");
                 if (method.DeprecatedIn is not null && engine.IsHardforkEnabled(method.DeprecatedIn.Value))
                     throw new InvalidOperationException($"Cannot call this method after hardfork {method.DeprecatedIn}.");
-                ExecutionContextState state = context.GetState<ExecutionContextState>();
-                if (!state.CallFlags.HasFlag(method.RequiredCallFlags))
+                var state = context.GetState<ExecutionContextState>();
+                if (!method.CheckCallFlags(engine, state.CallFlags))
```
Comment (roman-khimov) [inline src/Neo/SmartContract/Native/NativeContract.cs:387]: While this should work, I'm worried about the level of complexity and duplication we add here. Ideally I'd like to see proper per-hardfork manifests being used for invocations, this at least would eliminate the need for `RegisterAndUnregisterCallflags` (everything can be conveyed at `CreateMethodMetadataInternal` stage).

**Review by roman-khimov — COMMENTED** [review]
`vote` method requires this too.

File: src/Neo/SmartContract/Native/NativeContract.cs (hunk 1 above)
Comment (shargon) [reply src/Neo/SmartContract/Native/NativeContract.cs]: Trying a new way

File: src/Neo/ProtocolSettings.cs [hunk 2]
```diff
@@ -131,8 +146,8 @@ public record ProtocolSettings
         /// <returns>The loaded <see cref="ProtocolSettings"/>.</returns>
         public static ProtocolSettings Load(string path, bool optional = true)
         {
-            IConfigurationRoot config = new ConfigurationBuilder().AddJsonFile(path, optional).Build();
-            IConfigurationSection section = config.GetSection("ProtocolConfiguration");
+            var config = new ConfigurationBuilder().AddJsonFile(path, optional).Build();
+            var section = config.GetSection("ProtocolConfiguration");
```
Comment (roman-khimov) [inline src/Neo/ProtocolSettings.cs:150]: You can reuse `Load(stream)` here then.

File: src/Neo/ProtocolSettings.cs [hunk 3]
```diff
@@ -131,8 +146,8 @@ public record ProtocolSettings
         /// <returns>The loaded <see cref="ProtocolSettings"/>.</returns>
         public static ProtocolSettings Load(string path, bool optional = true)
         {
-            IConfigurationRoot config = new ConfigurationBuilder().AddJsonFile(path, optional).Build();
-            IConfigurationSection section = config.GetSection("ProtocolConfiguration");
+            var config = new ConfigurationBuilder().AddJsonFile(path, optional).Build();
```
Comment (cschuchardt88) [inline src/Neo/ProtocolSettings.cs:149]:
```suggestion
            using var config = new ConfigurationBuilder().AddJsonFile(path, optional).Build();
```

File: src/Neo/ProtocolSettings.cs [hunk 4]
```diff
@@ -123,6 +124,20 @@ public record ProtocolSettings
...
         public static ProtocolSettings Custom { get; set; }
 
+        /// <summary>
+        /// Loads the <see cref="ProtocolSettings"/> at the specified path.
+        /// </summary>
+        /// <param name="stream">The stream of the settings.</param>
+        /// <returns>The loaded <see cref="ProtocolSettings"/>.</returns>
+        public static ProtocolSettings Load(Stream stream)
+        {
+            var config = new ConfigurationBuilder().AddJsonStream(stream).Build();
```
Comment (cschuchardt88) [inline src/Neo/ProtocolSettings.cs:134]:
```suggestion
            using var config = new ConfigurationBuilder().AddJsonStream(stream).Build();
```

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]
Please fix these memory leaks

File: src/Neo/ProtocolSettings.cs [hunk 5]
```diff
@@ -131,7 +131,7 @@ public record ProtocolSettings
         /// <returns>The loaded <see cref="ProtocolSettings"/>.</returns>
         public static ProtocolSettings Load(Stream stream)
         {
-            var config = new ConfigurationBuilder().AddJsonStream(stream).Build();
+            using var config = new ConfigurationBuilder().AddJsonStream(stream).Build();
```
Comment (cschuchardt88) [inline src/Neo/ProtocolSettings.cs:134]: `ConfigurationBuilder` on build uses `ConfigurationRoot` may have to cast to it

https://github.com/dotnet/runtime/blob/9d5a6a9aa463d6d10b0b0ba6d5982cc82f363dc3/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationBuilder.cs#L53C24-L53C41

File: src/Neo/ProtocolSettings.cs [hunk 6]
```diff
@@ -123,6 +124,20 @@ public record ProtocolSettings
 
         public static ProtocolSettings Custom { get; set; }
 
+        /// <summary>
+        /// Loads the <see cref="ProtocolSettings"/> at the specified path.
```
[the remark this answers is not in the document]
Comment (Jim8y) [reply src/Neo/ProtocolSettings.cs:128]: for the specified path?

File: src/Neo/SmartContract/Native/CryptoLib.cs [hunk 7]
```diff
@@ -100,7 +100,7 @@ public static bool VerifyWithECDsa(byte[] message, byte[] pubkey, byte[] signatu
         }
 
         // This is for solving the hardfork issue in https://github.com/neo-project/neo/pull/3209
-        [ContractMethod(true, Hardfork.HF_Cockatrice, CpuFee = 1 << 15, Name = "verifyWithECDsa")]
+        [ContractMethod(true, Hardfork.HF_Cockatrice, CpuFee = 1 << 15, Name = nameof(VerifyWithECDsa))]
```
Comment (Jim8y) [inline src/Neo/SmartContract/Native/CryptoLib.cs:103]: i have a question that is not related to this pr, can we rename the CPUFee to system fee?

Comment (roman-khimov) [reply src/Neo/SmartContract/Native/CryptoLib.cs:103]: No. There is `CpuFee`. And there is `StorageFee`. They're very different even though system fee pays for everything in the end.

File: src/Neo/ProtocolSettings.cs (hunk 5 above)
Comment (shargon) [reply src/Neo/ProtocolSettings.cs:134]: It was not a memory leak...

**Comment by shargon** [comment]
Ready to review again

File: src/Neo/ProtocolSettings.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (shargon) [reply src/Neo/ProtocolSettings.cs]: fixed

File: src/Neo/ProtocolSettings.cs (hunk 2 above)
Comment (shargon) [reply src/Neo/ProtocolSettings.cs]: Fixed

File: src/Neo/ProtocolSettings.cs (hunk 5 above)
Comment (cschuchardt88) [reply src/Neo/ProtocolSettings.cs]: How isn't it a memory leak, if the memory for the stream is the still allocated?
https://github.com/dotnet/runtime/blob/9d5a6a9aa463d6d10b0b0ba6d5982cc82f363dc3/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationRoot.cs#L99C9-L112C10

File: src/Neo/ProtocolSettings.cs [hunk 8]
```diff
@@ -146,11 +146,21 @@ public static ProtocolSettings Load(Stream stream)
...
         public static ProtocolSettings Load(string path, bool optional = true)
         {
-            using var config = new ConfigurationBuilder().AddJsonFile(path, optional).Build();
-            var section = config.GetSection("ProtocolConfiguration");
-            var settings = Load(section);
-            CheckingHardfork(settings);
-            return settings;
+            if (!File.Exists(path))
+            {
+                if (optional)
```
Comment (cschuchardt88) [inline src/Neo/ProtocolSettings.cs:150]: What is the point of optional here? Just use what you had before. Should be renamed to properly identify `optional`, something along the lines of `useDefaultOnNonExists`. Nor even the comments discard this.

Comment (shargon) [reply src/Neo/ProtocolSettings.cs:150]: Preserve the behavior, if it's my decision I remove this arg

**Review by roman-khimov — APPROVED** [review]

File: src/Neo/ProtocolSettings.cs (hunk 8 above)
Comment (cschuchardt88) [reply src/Neo/ProtocolSettings.cs:150]: than remove it.

Comment (shargon) [reply src/Neo/ProtocolSettings.cs:150]: In a different PR

File: src/Neo/SmartContract/Native/NativeContract.cs [hunk 9]
```diff
@@ -140,11 +140,12 @@ protected NativeContract()
             // Reflection to get the methods
 
             List<ContractMethodMetadata> listMethods = [];
-            foreach (MemberInfo member in GetType().GetMembers(BindingFlags.Instance | BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public))
+            foreach (var member in GetType().GetMembers(BindingFlags.Instance | BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public))
             {
-                ContractMethodAttribute attribute = member.GetCustomAttribute<ContractMethodAttribute>();
-                if (attribute is null) continue;
-                listMethods.Add(new ContractMethodMetadata(member, attribute));
+                foreach (var attribute in member.GetCustomAttributes<ContractMethodAttribute>())
```
Comment (AnnaShaleva) [inline src/Neo/SmartContract/Native/NativeContract.cs:145]: Previously there was a check for `if (attribute is null)`, is it possible that attribute is nil with the current implementation? I think it's OK, but we'd better make sure.

Comment (shargon) [reply src/Neo/SmartContract/Native/NativeContract.cs:145]: No, now can't be null

**Review by AnnaShaleva — APPROVED** [review]

**Review by Jim8y — APPROVED** [review]

**Review by roman-khimov — APPROVED** [review]
